base/shell/explorer/desktop.cpp

```diff
@@ -22,104 +22,184 @@
 
 class CDesktopThread
 {
-    HANDLE m_hEvent;
+private:
     CComPtr<ITrayWindow> m_Tray;
+    HANDLE m_hEvent;
+    HANDLE m_hThread;
 
-    DWORD DesktopThreadProc()
-    {
-        CComPtr<IShellDesktopTray> pSdt;
-        HANDLE hDesktop;
-        HRESULT hRet;
-
-        OleInitialize(NULL);
+    DWORD DesktopThreadProc();
+    static DWORD WINAPI s_DesktopThreadProc(LPVOID lpParameter);
+    
+public:
+    CDesktopThread();
+    virtual ~CDesktopThread();
 
-        hRet = m_Tray->QueryInterface(IID_PPV_ARG(IShellDesktopTray, &pSdt));
-        if (!SUCCEEDED(hRet))
-            return 1;
+    HRESULT Initialize(ITrayWindow *pTray);
+    void Destroy();
+};
 
-        hDesktop = _SHCreateDesktop(pSdt);
-        if (hDesktop == NULL)
-            return 1;
+/*******************************************************************/
 
-        if (!SetEvent(m_hEvent))
-        {
-            /* Failed to notify that we initialized successfully, kill ourselves
-            to make the main thread wake up! */
-            return 1;
-        }
-
-        _SHDesktopMessageLoop(hDesktop);
+CDesktopThread::CDesktopThread() :
+    m_Tray(NULL),
+    m_hEvent(NULL),
+    m_hThread(NULL)
+{
+}
 
-        /* FIXME: Properly rundown the main thread! */
-        ExitProcess(0);
+CDesktopThread::~CDesktopThread()
+{
+    Destroy();
+}
 
-        return 0;
+HRESULT CDesktopThread::Initialize(ITrayWindow *pTray)
+{
+    HANDLE hEvent;
+    HANDLE hThread;
+    HANDLE Handles[2];
+    
+    if (!pTray)
+    {
+        return E_FAIL;
     }
-
-    static DWORD CALLBACK s_DesktopThreadProc(IN OUT LPVOID lpParameter)
+    
+    if (m_Tray || m_hEvent || m_hThread)
     {
-        return reinterpret_cast<CDesktopThread*>(lpParameter)->DesktopThreadProc();
+        return E_FAIL;
     }
+    
+    hEvent = CreateEventW(NULL, FALSE, FALSE, NULL);
 
-public:
-    CDesktopThread() :
-        m_hEvent(NULL),
-        m_Tray(NULL)
+    if (!hEvent)
     {
+        return E_FAIL;
     }
 
-    HRESULT Initialize(IN OUT ITrayWindow *pTray)
-    {
-        HANDLE hThread;
-        HANDLE Handles[2];
+    hThread = CreateThread(NULL, 0, s_DesktopThreadProc, (LPVOID)this, CREATE_SUSPENDED, NULL);
```

Comment:
What is the purpose of creating it suspended and then resuming it?

Reply:
To set `m_Tray` and `m_hEvent` value before the thread executes.

Reply:
but you can do that by just assigning them before CreateThread?

Reply:
Fixed.